module/zfs/dmu_send.c

```diff
@@ -2536,19 +2554,27 @@ dmu_send_impl(struct dmu_send_params *dspp)
 		goto out;
 	}
 
-	bzero(drr, sizeof (dmu_replay_record_t));
-	drr->drr_type = DRR_END;
-	drr->drr_u.drr_end.drr_checksum = dsc.dsc_zc;
-	drr->drr_u.drr_end.drr_toguid = dsc.dsc_toguid;
+	/*
+	 * Send the DRR_END record if this is not a partial stream.
+	 * Otherwise, the omitted DRR_END record will signal to
+	 * the receive side that the stream is incomplete.
+	 */
+	if (!dspp->partialok) {
```

Comment:
Given this behavior, the flag isn't really "partial OK", it's "send the partially received state".  You wouldn't want to use it on a fully-received snapshot (or filesystem).  So I think the command should fail if there is no partially received state.

Reply:
Would looking up the RESUME_TOGUID and RESUME_TONAME in dmu_send (if partialok), and failing if not found cover this?

Reply:
That sounds right to me.